### DISCUSSION TITLE
Additional parameters for `meal-plan` API

### DISCUSSION THREAD

**Discussion by WoosterInitiative** [TandoorRecipes/recipes#2597] [Ideas]
I'm trying to build a "block" for my DAKBoard dashboard that pulls in our meal plans. However, none of the URLs give me the dynamic filter that I need for that nor provide sorting.

I suggest adding the following functionality (I've already toyed with it in a local environment, the changes to the code are actually fairly limited, I just haven't gotten a functional test environment setup to test what I've done... yet):

1. Add the ability to use "today" in the `from_date` and `to_date` parameters
2. Add an optional `offset` parameter that can be used with *either* `from_date` or `to_date` to create a "dynamic" range
3. Add an optional `ordering` parameter that allows for django-like sorting using a property of the list (i.e. "date" or "-date" for descending sorting)

Again, all of this is remarkably easy using django's querysets, just a few tricks to do the math properly and do a little error checking.

Thoughts? I've also thought that there might be a use for this kind of tool as a mixin to allow for similar filtering and sorting of other endpoints, but this is the one I'm needing right now.

```
class MealPlanViewSet(viewsets.ModelViewSet):
    """
    list:
    optional parameters

    - **from_date**: filter from (inclusive) a certain date onward (e.g. 2000-01-01
      **OR** "yesterday", "today", "tomorrow", "0" (today), "7" (in 7 days), etc.)
    - **to_date**: filter upward to (inclusive) certain date (e.g. 2300-01-01 **OR**
      "yesterday", "today", "tomorrow", "0" (today), "7" (in 7 days), etc.)
    - "date_offset": number of days to offset **from_date** or **to_date**; "0" for same
      day, negative values allowed
    - **ordering**: which field to order the values by (e.g. 'date'); use a "-" to order
      descending (e.g. '-date')

    """

    queryset = MealPlan.objects
    serializer_class = MealPlanSerializer
    permission_classes = [
        (CustomIsOwner | CustomIsShared) & CustomTokenHasReadWriteScope
    ]

    def get_queryset(self):
        queryset = (
            self.queryset.filter(
                Q(created_by=self.request.user) | Q(shared=self.request.user)
            )
            .filter(space=self.request.space)
            .distinct()
            .all()
        )

        from_date = self.request.query_params.get("from_date", None)
        to_date = self.request.query_params.get("to_date", None)
        offset = self.request.query_params.get("offset", None)
        ordering = self.request.query_params.get("ordering", None)

        if all([from_date, to_date, offset]):
            raise APIException(
                code=400,
                detail="No more than two of 'from_date', 'to_date', and 'offset' allowed.",
            )

        # set date range if `from_date` specified with `offset`
        if offset and from_date:
            ranged_date = from_date + timedelta(days=offset)
            if offset > 0:
                to_date = ranged_date
            elif offset < 0:
                to_date = from_date
                from_date = ranged_date

        # set date range if `to_date` specified with `offset`
        if offset and to_date:
            ranged_date = to_date + timedelta(days=offset)
            if offset > 0:
                from_date = to_date
                to_date = ranged_date
            if offset < 0:
                from_date = ranged_date

        if from_date:
            from_date = make_date(from_date)
            queryset = queryset.filter(date__gte=from_date)

        if to_date:
            to_date = make_date(to_date)
            queryset = queryset.filter(date__lte=to_date)

        if ordering:
            queryset = queryset.order_by(ordering)

        return queryset
```

`make_date` is a "helper" that tries to take an input and either check if it's a date or convert an integer or "today" into a date.

I noticed while reviewing this that the code is a little out of order, but I hope that the concept is clear. Once I have a dev environment setup, I'll work on the PR some more.

**Comment by smilerz** [comment]
Please review the QueryParam class that is implemented on some of the other APIs (RecipeViewset as example)

There is also another feature request somewhere to allow filtering on mealplan type.

I'll let vabene1111 make the final call - but current standard is to have a unique parameter for each sort type named explicitly (date, created_by, created_on, etc) so that they can be applied cumulatively.  We should probably implement API sorting consistently across models.

I'm not a huge fan of using strings and numbers in the to_date/from_date - it overlaps with offset in ways that aren't intuitive.  I'd stick with special words - explicitly defined in the QueryParam help text.  Then leave integer values to the offset.

**Reply by WoosterInitiative** [reply]
I'll take a look at that. I tried looking through the code but didn't see anything.

Those suggestions make a lot of sense. I'll continue working on a local copy of it and wait to submit a PR until vabene1111 has provided any input.

**Reply by vabene1111** [reply]
i agree with what smilierz said, please orient yourself by looking at other classes, mainly the recipe one for how parameters are named and implemented. Especially having ?today=true instead of from_date=today is something i would prefer. 

If you need help setting up your dev environment please let me know, also it would be great if we could add some tests to the endpoint because once you start adding many filters tests help not breaking one if you work on the other.

**Reply by vabene1111** [reply]
one additonal note, i am currently playing around with changing the MealPlan entry to allow a from_date and to_date for each entry for multi day planning, just so you know that you might need to change a few things on your code after that. But not sure yet when/how I will do it.

**Reply by WoosterInitiative** [reply]
@vabene1111 how can I contact you about setting up my dev environment? I'm having issues getting Vue working, which for my purposes, I only need so I can make an API token. I'm having trouble in Windows, WSL, and a full Ubuntu desktop and I just don't know enough about node to debug my issue.

Also, have you looked at Dev Containers at all? I'm no expert, but the idea seems good. If there's any interest at all, I'll start a separate discussion on it.

**Reply by smilerz** [reply]
@WoosterInitiative join the discord server and i can try to help you

**Reply by vabene1111** [reply]
Or you could just create a token in the db if you want to save some time. Just make sure it has read, write as a scope and the token can basically be whatever you want
